File: core/src/main/java/com/cv4j/netdiscovery/core/domain/HttpRequestBody.java
package com.cv4j.netdiscovery.core.domain;

import com.alibaba.fastjson.JSONObject;
import lombok.Getter;
import lombok.Setter;
import org.apache.http.NameValuePair;
import org.apache.http.client.utils.URLEncodedUtils;
import org.apache.http.message.BasicNameValuePair;

import java.io.Serializable;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created by tony on 2018/2/19.
 */
public class HttpRequestBody implements Serializable{

    public static abstract class ContentType {

        public static final String JSON = "application/json";

        public static final String FORM = "application/x-www-form-urlencoded";

        public static final String MULTIPART = "multipart/form-data";
    }

    @Setter
    @Getter
    private byte[] body;

    @Setter
    @Getter
    private String contentType;

    @Setter
    @Getter
    private String encoding;

    public HttpRequestBody(byte[] body, String contentType, String encoding) {
        this.body = body;
        this.contentType = contentType;
        this.encoding = encoding;
    }

    /**
     * post body使用jsonObject时，传递一个jsonObject即可生成HttpRequestBody对象
     * @param jsonObject
     * @return
     */
    public static HttpRequestBody json(JSONObject jsonObject) {

        return json(jsonObject.toJSONString());
    }

    public static HttpRequestBody json(String json) {

        return json(json,"UTF-8");
    }

    public static HttpRequestBody json(String json, String encoding) {

        try {
            return new HttpRequestBody(json.getBytes(encoding), ContentType.JSON, encoding);
        } catch (UnsupportedEncodingException e) {
            throw new IllegalArgumentException("illegal encoding " + encoding, e);
        }
    }

    public static HttpRequestBody custom(byte[] body, String contentType, String encoding) {
        return new HttpRequestBody(body, contentType, encoding);
    }

    public static HttpRequestBody form(Map<String,Object> params, String encoding){

        List<NameValuePair> nameValuePairs = new ArrayList<>(params.size());
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            nameValuePairs.add(new BasicNameValuePair(entry.getKey(), String.valueOf(entry.getValue())));
        }
        try {
            return new HttpRequestBody(URLEncodedUtils.format(nameValuePairs, encoding).getBytes(encoding), ContentType.FORM, encoding);
        } catch (UnsupportedEncodingException e) {
            throw new IllegalArgumentException("illegal encoding " + encoding, e);
        }
    }

}

File: core/core/src/main/kotlin/cn/netdiscovery/core/domain/HttpRequestBody.kt
package cn.netdiscovery.core.domain

import com.google.gson.JsonObject
import org.apache.http.NameValuePair
import org.apache.http.client.utils.URLEncodedUtils
import org.apache.http.message.BasicNameValuePair
import java.io.Serializable
import java.io.UnsupportedEncodingException
import java.util.*

/**
 *
 * @FileName:
 *          cn.netdiscovery.core.domain.HttpRequestBody
 * @author: Tony Shen
 * @date: 2020-10-10 16:11
 * @version: V1.0 <描述当前版本功能>
 */
class HttpRequestBody(body: ByteArray?, contentType: String?, encoding: String?) :
    Serializable {

    var body: ByteArray?

    var contentType: String?

    var encoding: String?

    init {
        this.body = body
        this.contentType = contentType
        this.encoding = encoding
    }

    companion object {

        /**
         * post body 使用 jsonObject 时，传递一个 jsonObject 即可生成 HttpRequestBody 对象
         * @param jsonObject
         * @return
         */
        @JvmStatic
        fun json(jsonObject: JsonObject): HttpRequestBody {
            return json(jsonObject.toString())
        }

        @JvmOverloads
        fun json(json: String, encoding: String = "UTF-8"): HttpRequestBody {
            return try {
                HttpRequestBody(json.toByteArray(charset(encoding)), ContentType.JSON, encoding)
            } catch (e: UnsupportedEncodingException) {
                throw IllegalArgumentException("illegal encoding $encoding", e)
            }
        }

        fun custom(body: ByteArray?, contentType: String?, encoding: String?): HttpRequestBody {
            return HttpRequestBody(body, contentType, encoding)
        }

        fun form(params: Map<String?, Any>, encoding: String): HttpRequestBody {
            val nameValuePairs: MutableList<NameValuePair> = ArrayList(params.size)
            for ((key, value) in params) {
                nameValuePairs.add(BasicNameValuePair(key, value.toString()))
            }
            return try {
                HttpRequestBody(
                    URLEncodedUtils.format(nameValuePairs, encoding).toByteArray(charset(encoding)),
                    ContentType.FORM,
                    encoding
                )
            } catch (e: UnsupportedEncodingException) {
                throw IllegalArgumentException("illegal encoding $encoding", e)
            }
        }
    }

    object ContentType {
        const val JSON = "application/json"
        const val FORM = "application/x-www-form-urlencoded"
        const val MULTIPART = "multipart/form-data"
    }
}